### PR TITLE
GEOMESA-2158 Add a geomesa console command

### PR DESCRIPTION
**Pull request by aheyne**
Implement a 'bin/geomesa-* console' command that starts a scala console with the geomesa classpath and config loaded.

Signed-off-by: Austin Heyne <aheyne@ccri.com>

### CODE REVIEW

File: geomesa-tools/bin/common-functions.sh [hunk 1]
```diff
@@ -279,6 +279,49 @@ function registerAutocomplete() {
...
+  classpath=${1}
+  # Check if we already downloaded scala
+  if [[ -d "${%%gmtools.dist.name%%_HOME}/dist/scala-%%scala.version%%/" ]]; then
+    scalaCMD="${%%gmtools.dist.name%%_HOME}/dist/scala-%%scala.version%%/bin/scala"
+  else
+    if which scala > /dev/null; then
+      scalaCMD=`which scala`
+      version=$(scala -version 2>&1 | grep %%scala.binary.version%%)
+      if [[ -z "${version}" ]]; then
+        read -p "The wrong Scala version is installed, do you want to download the correct one? (This will not effect your current install) Y\n" -n 1 -r
```
Comment (elahrvivaz) [inline geomesa-tools/bin/common-functions.sh]: grammar: effect->affect

File: geomesa-tools/bin/common-functions.sh [hunk 2]
```diff
@@ -279,6 +279,49 @@ function registerAutocomplete() {
   fi
 }
 
+function installScala() {
+  # Download scala to %%gmtools.dist.name%%_HOME}/dist/
+  sourceURL="https://downloads.lightbend.com/scala/%%scala.version%%/scala-%%scala.version%%.tgz"
+  outputFile="${%%gmtools.dist.name%%_HOME}/dist/scala-%%scala.version%%.tgz"
+  wget $sourceURL -O $outputFile
```
Comment (elahrvivaz) [inline geomesa-tools/bin/common-functions.sh]: prefer `curl` over `wget`, as it's more widely installed. you can just call the 'downloadUrls' function, which will take care of it for you. it has a confirmation prompt, so you'll want to check that the file was actually downloaded before continuing. the logic below will need to be tweaked slightly, so you don't have double confirmations.

Comment (jahhulbert-ccri) [reply geomesa-tools/bin/common-functions.sh]: +1

**Comment by jahhulbert-ccri** [comment]
one thought...since Java9 will have a Java Console what if we made this ``bin/geomesa scala-console`` so that later we could add ``bin/geomesa java-console`` it's a thought.

File: geomesa-tools/bin/common-functions.sh (hunk 2 above)
Comment (aheyne) [reply geomesa-tools/bin/common-functions.sh]: Good Idea, I didn't notice that function.

**Comment by elahrvivaz** [comment]
@jahhulbert-ccri sounds good to me

**Comment by aheyne** [comment]
@jahhulbert-ccri Seems reasonable, I could see us having a python console at some point too.

File: geomesa-tools/bin/common-functions.sh [hunk 3]
```diff
@@ -279,6 +289,57 @@ function registerAutocomplete() {
   fi
 }
 
+function installScala() {
```
Comment (elahrvivaz) [inline geomesa-tools/bin/common-functions.sh]: this function can be removed now i think?

Comment (aheyne) [reply geomesa-tools/bin/common-functions.sh]: Woops, I thought I removed that.

**Comment by elahrvivaz** [comment]
@aheyne could you add something in the docs about this? Would be awesome to have a snippet about it in the quick start section too, something like copy out one of our basic unit tests (create schema, write features, read back features) as something that you can easily run in the console to get started.

**Comment by jahhulbert-ccri** [comment]
i'm reviewing and will merge if it looks good! good idea @aheyne should be very useful.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/status/ScalaConsoleCommand.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.tools.status
+
+import com.beust.jcommander.Parameters
+import org.locationtech.geomesa.tools.Command
+
+/**
+ * Note: this class is a placeholder for the 'console' function implemented in the 'geomesa-*' script, to get it
```
Comment (aheyne) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/status/ScalaConsoleCommand.scala]: Note to me, fix this.

File: docs/user/cli/console.rst [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            <script src="https://asciinema.org/a/162999.js" id="asciicast-162999"  async></script>
+
+    .. tab:: Transcript
+
+        First we run the command, this starts up the Scala REPL with all of the
+        resources we need.
+
+        .. code-block:: scala
+
+            aheyne@aheyne:/tmp/geomesa-fs_2.11-2.0.0-SNAPSHOT$ bin/geomesa-fs scala-console
```
Comment (elahrvivaz) [inline docs/user/cli/console.rst]: could you change the `aheyne@aheyne:` to just be `$`? that's what we do other places. Also, do you think it makes sense to pull the command up into a distinct bash code block, then use scala for the output?

File: docs/user/cli/console.rst [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        .. code-block:: scala
+
+            scala> while (reader.hasNext()) { println(reader.next().toString()) }
+            ScalaSimpleFeature:719024956:719024956|||GANG||120|6|1|6|1|Brazil|BR|Sun Dec 31 19:00:00 EST 2017|POINT (-55 -10)
+            ScalaSimpleFeature:719024898:719024898|||SYDNEY|AUS|010|14|2|14|4|Sydney, New South Wales, Australia|AS|Sun Dec 31 19:00:00 EST 2017|POINT (151.217 -33.8833)
+            ScalaSimpleFeature:719024882:719024882|SECURITY COUNCIL||PYONGYANG|PRK|163|2|1|2|1|Russia|RS|Sun Dec 24 19:00:00 EST 2017|POINT (100 60)
+            ScalaSimpleFeature:719024881:719024881|||RUSSIA|RUS|042|2|1|2|3|Allegheny County, Pennsylvania, United States|US|Sun Dec 24 19:00:00 EST 2017|POINT (-80.1251 40.6253)
+            ScalaSimpleFeature:719025149:719025149|ARGENTINE|ARG|DIOCESE||010|1|1|1|4|Corrientes, Corrientes, Argentina|AR|Sun Dec 31 19:00:00 EST 2017|POINT (-58.8341 -27.4806)
+            ...
+
```
Comment (elahrvivaz) [inline docs/user/cli/console.rst]: could you add a close on the reader and a dispose on the datastore? while probably not strictly necessary, it's nice to encourage good habits :)

File: geomesa-tools/bin/common-functions.sh [hunk 7]
```diff
@@ -52,26 +52,36 @@ function findJars() {
...
-# 
+#  - --noPrompt [optional] bypass confirmation. Useful when
+#    you'd like to provide your own confirmation
+#    messages/handling
+#
 # for example:
 #   urls=("1" "2" "3")
 #   downloadURLS /tmp/foobar urls[@]
+# or
+#   downloadURLS /tmp/foobar urls[@] --noPrompt
```
Comment (elahrvivaz) [inline geomesa-tools/bin/common-functions.sh]: could we change ``--noPrompt`` to ``--no-prompt``? seems more consistent...

File: docs/user/cli/console.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+download and run the appropriate version Scala if it is not available. This is a
+portable install and will not alter the machines current configuration.
+
+Example usage
+-------------
+
+The following example was performed against the FileSystem Datastore generated by the
+:ref:`fsds_quickstart`. It shows how to use the ``scala-console`` command to connect
+to a FileSystem Datastore, discover the feature type names, get the schema and query for data.
+
```
Comment (elahrvivaz) [inline docs/user/cli/console.rst]: Should probably add a `.. note::` here to emphasize that this is running against the output of the quick start

File: docs/user/cli/console.rst (hunk 5 above)
Comment (aheyne) [reply docs/user/cli/console.rst]: Yeah, I updated the asciinema to only be `$` as well.

File: geomesa-tools/bin/common-functions.sh (hunk 7 above)
Comment (aheyne) [reply geomesa-tools/bin/common-functions.sh]: Changed

File: docs/user/cli/console.rst (hunk 6 above)
Comment (aheyne) [reply docs/user/cli/console.rst]: Sure

File: docs/user/cli/console.rst (hunk 8 above)
Comment (aheyne) [reply docs/user/cli/console.rst]: Done

**Comment by elahrvivaz** [comment]
@aheyne will that ascii video stay up indefinitely?

**Pull request merged**